src/map/instance.cpp

```diff
@@ -339,314 +529,300 @@ static int instance_addnpc_sub(struct block_list *bl, va_list ap)
+	std::shared_ptr<s_instance_db> db = instance_search_db_name(name);
+
+	if (!db) {
+		ShowError("instance_create: Unknown instance %s creation was attempted.\n", name);
+		return -1;
+	}
+
+	struct map_session_data *sd;
+	struct party_data *pd;
+	struct guild *gd;
+	struct clan* cd;
+	int instance_id;
 
 	switch(mode) {
 		case IM_NONE:
 			break;
 		case IM_CHAR:
-			if ((sd = map_charid2sd(owner_id)) == NULL) {
-				ShowError("instance_create: character %d not found for instance '%s'.\n", owner_id, name);
+			if (!(sd = map_charid2sd(owner_id))) {
+				ShowError("instance_create: Character %d not found for instance '%s'.\n", owner_id, name);
 				return -2;
 			}
 			if (sd->instance_id)
 				return -3; // Player already instancing
 			break;
 		case IM_PARTY:
-			if ((pd = party_search(owner_id)) == NULL) {
-				ShowError("instance_create: party %d not found for instance '%s'.\n", owner_id, name);
+			if (!(pd = party_search(owner_id))) {
+				ShowError("instance_create: Party %d not found for instance '%s'.\n", owner_id, name);
 				return -2;
 			}
 			if (pd->instance_id)
 				return -3; // Party already instancing
 			break;
 		case IM_GUILD:
-			if ((gd = guild_search(owner_id)) == NULL) {
-				ShowError("instance_create: guild %d not found for instance '%s'.\n", owner_id, name);
+			if (!(gd = guild_search(owner_id))) {
+				ShowError("instance_create: Guild %d not found for instance '%s'.\n", owner_id, name);
 				return -2;
 			}
 			if (gd->instance_id)
 				return -3; // Guild already instancing
 			break;
 		case IM_CLAN:
-			if ((cd = clan_search(owner_id)) == NULL) {
-				ShowError("instance_create: clan %d not found for instance '%s'.\n", owner_id, name);
+			if (!(cd = clan_search(owner_id))) {
+				ShowError("instance_create: Clan %d not found for instance '%s'.\n", owner_id, name);
 				return -2;
 			}
 			if (cd->instance_id)
 				return -3; // Clan already instancing
 			break;
 		default:
-			ShowError("instance_create: unknown mode %u for owner_id %d and name %s.\n", mode, owner_id, name);
+			ShowError("instance_create: Unknown mode %u for owner_id %d and name %s.\n", mode, owner_id, name);
 			return -2;
 	}
 
-	// Searching a Free Instance
-	// 0 is ignored as this means "no instance" on maps
-	ARR_FIND(1, MAX_INSTANCE_DATA, i, instance_data[i].state == INSTANCE_FREE);
-	if( i >= MAX_INSTANCE_DATA )
+	if (instances.size() > MAX_INSTANCE_DATA)
 		return -4;
 
-	instance_data[i].type = db->id;
-	instance_data[i].state = INSTANCE_IDLE;
-	instance_data[i].owner_id = owner_id;
-	instance_data[i].mode = mode;
-	instance_data[i].keep_limit = 0;
-	instance_data[i].keep_timer = INVALID_TIMER;
-	instance_data[i].idle_limit = 0;
-	instance_data[i].idle_timer = INVALID_TIMER;
-	instance_data[i].regs.vars = i64db_alloc(DB_OPT_RELEASE_DATA);
-	instance_data[i].regs.arrays = NULL;
-	instance_data[i].cnt_map = 0;
+	instance_id = static_cast<int>(instances.size()) + 1;
```

Comment:
The instance ID isn't unique using this method. For example instances 1, 2, 3 are created, instance 1 expire, the next ID will be 3